sql/src/main/java/io/crate/operation/reference/sys/check/cluster/TablesNeedUpgradeSysCheck.java

```diff
@@ -0,0 +1,163 @@
+                                       org.apache.lucene.util.Version.LATEST.toString() + "'";
+    private static final String STMT_NOT_IN = " AND schema_name || '.' || table_name != ANY (?)";
+    private static final String STMT_SUFFIX = " order by 1";
+    private static final int LIMIT = 50_000;
+
+    private final SQLOperations.Session session;
+    private final ClusterService clusterService;
+    private volatile Collection<String> tablesNeedUpgrade;
+
+    @Inject
+    public TablesNeedUpgradeSysCheck(ClusterService clusterService, SQLOperations sqlOperations) {
+        super(ID, DESCRIPTION, Severity.MEDIUM);
+        this.clusterService = clusterService;
+        this.session = sqlOperations.createSession("sys", Option.NONE, LIMIT);
+        this.session.parse(SQLOperations.Session.UNNAMED, STMT, Collections.emptyList());
+        sqlOperations.createSession(null, Option.NONE, 1);
+    }
+
+    @Override
+    public BytesRef description() {
+        String linkedDescriptionBuilder = DESCRIPTION + tablesNeedUpgrade + ' ' + LINK_PATTERN + ID;
+        return new BytesRef(linkedDescriptionBuilder);
+    }
+
+    @Override
+    public CompletableFuture<?> computeResult() {
+        Collection<String> tablesNeedReindexing =
+            LuceneVersionChecks.tablesNeedReindexing(clusterService.state().metaData());
+        String statement = createSQLStatement(tablesNeedReindexing);
+
+        final CompletableFuture<Collection<String>> result = new CompletableFuture<>();
+        try {
+            session.parse(SQLOperations.Session.UNNAMED, statement, Collections.emptyList());
+            session.bind(
+                SQLOperations.Session.UNNAMED,
+                SQLOperations.Session.UNNAMED,
+                tablesNeedReindexing.isEmpty() ?
+                    Collections.emptyList() :
+                    Collections.singletonList(tablesNeedReindexing.toArray(new String[] {})),
+                null);
+            session.execute(
+                SQLOperations.Session.UNNAMED,
+                0,
+                new SycCheckResultReceiver(result));
+            session.sync();
+        } catch (Throwable t) {
+            result.completeExceptionally(t);
+        }
+        return result.handle((tableNames, throwable) -> {
+            if (throwable == null) {
+                tablesNeedUpgrade = tableNames;
+            } else {
+                LOGGER.error("error while checking for tables that need upgrade", throwable);
+            }
+            return CompletableFuture.completedFuture(null);
+        });
+    }
+
+    @Override
+    public boolean validate() {
+        return tablesNeedUpgrade == null || tablesNeedUpgrade.isEmpty();
+    }
+
+    private static String createSQLStatement(Collection<String> tablesNeedReindexing) {
+        String statement;
+        if (tablesNeedReindexing.isEmpty()) {
+            statement = STMT + STMT_SUFFIX;
+        } else {
+            statement = STMT + STMT_NOT_IN + STMT_SUFFIX;
```

Comment:
Isn't it possible to always use the same statement?  x != ANY([]) matches everything.

If not, the upfront parse call on the session doesn't really make any sence anyway since it's a statement that's never executed